### ISSUE TITLE
Keyboard input laggy after installing tuxedo control center on arch

### ISSUE THREAD

**Issue by func0der** [tuxedocomputers/tuxedo-control-center#262]
Hey there,

I am installing the Tuxedo Control center from the AUR repository.
Using version 1.2.4

As soon as the package is installed, the keyboard input gets laggy and pretty regularily ignores complete inputs.

I am not sure, but it may aligns with the spikes of the `tccd` "spiking" at around 2-3 % cpu load. This happens every ~4 seconds or so. Hard to say from a quick look at htop.

I can not find any unusual activity in the kernel log. I can not find any log for tccd, so I don't know if there is something in there.

I am also unsure why this even happens for a control software for cpu frequencies.
I imagine, that it could somehow regulate the cpu down to a very low frequency and this makes the whole system lag. But it seems  not affect the rest of the system. At least not that I can see it.

Any ideas?

**Comment by func0der** [comment]
Seems to have been introduced in 1.2.0
1.1.8 has no lag issues.

**Comment by func0der** [comment]
First bad commit: https://github.com/tuxedocomputers/tuxedo-control-center/commit/9816af3

**Comment by func0der** [comment]
Something in the FanControlWorker makes it slow. If you disable the worker, no hang ups.

**Comment by func0der** [comment]
It seems that any ioAPI operation with the fans cause the pc to slow down
I can not pin down, why i started happening in the above mentioned commit. 

It appears to have nothing to do with the changes in the native lib part, since I reverted those back to "HEAD^" and it still happened.
I hope someone smarter than me can figure this out.

**Comment by mbway** [comment]
I think this is the same issue that my TUXEDO-Pulse-14-Gen1 has had for years.
The other day I upgraded from Ubuntu 20.04 to 22.04 and noticed that the keyboard was much more responsive. After installing tuxedo control center the laggyness returned so I investigated. the most annoying thing is when a key gets 'stuck' and repeats many times.

If I add `blacklist tuxedo_io` to `/etc/modprobe.d/blacklist.conf` the problem goes away.
I'm currently using Tuxedo control center 1.2.4 from <https://deb.tuxedocomputers.com/ubuntu>

I also cannot see anything useful in journalctl. I'm running the 5.15.0-58-generic kernel. I also do not see the lag spikes affect the system other than the keyboard responsiveness.

here are some examples of me typing (asdflkjh over and over) the same way with and without the module blacklisted. The repeated keys are *not* because I am holding that key down! I'm amazed I put up with it for so long. 

[unblacklisted.webm](https://user-images.githubusercontent.com/4923501/215267639-1c5426d1-6685-4aef-acfc-6bc5277374b0.webm)

[blacklisted.webm](https://user-images.githubusercontent.com/4923501/215267649-bbd3bba6-8119-43e5-8787-9757217816da.webm)

**Comment by xPathin** [comment]
Thank you all, this issue was driving me crazy, I tested so many things but couldn't find out that it was tccd causing the issues. It's even worse, it's not only laggy. It's also eating inputs.

I can confirm this problem existing in tuxedo control center 1.2.4 on Arch.

**Comment by mbway** [comment]
yes I think some inputs were missed on my machine as well. I used to think it was just the keyboard itself (maybe I wasn't pressing the keys hard enough etc)

**Comment by tuxedoxt** [comment]
Hello,

I had an initial look at the issue. With Fusion and our normal installation the keyboard was smooth.

With Pulse 14/15 Gen 1, however, it's noticable. I haven't verified it but it's likely the fan control/sensor readings is involved since it's done regularly and reads from the hardware IO (EC). Why this works different on these devices is another question.

**Comment by crisidev** [comment]
I can confirm it is pretty noticeable on Pulse 14 Gen 1.

**Comment by xPathin** [comment]
> Hello,
> 
> I had an initial look at the issue. With Fusion and our normal installation the keyboard was smooth.
> 
> With Pulse 14/15 Gen 1, however, it's noticable. I haven't verified it but it's likely the fan control/sensor readings is involved since it's done regularly and reads from the hardware IO (EC). Why this works different on these devices is another question.

Just FYI, because I have forgotten to include it in my comment, for me the issue exists on a TUXEDO Stellaris Intel Gen3 running Arch Linux and Wayland.

**Comment by vmaurin** [comment]
Same here with a Pulse 15 + Arch + Wayland. I had a linked issue open on tuxedo-keyboard project, but I pinpoint it to tccd too

**Comment by Donearm** [comment]
Can confirm too on Pulse 15 gen 1, Archlinux and both Wayland/Xorg. Blacklisting `tuxedo_io` module does help but it is still not responsive as it was months ago.

**Comment by alexbradd** [comment]
Can confirm too on Pulse 15 gen 2. In my case it is especially bad with the arrow keys: it continues to read repeated inputs for half a seconds after release in the worst cases. Running Arch with kernel 6.1.10 and tcc 1.2.4-1 (from AUR).

**Comment by matt-hires** [comment]
Same problem here with the following configuration:

```
System:
  Host: /*removed*/ Kernel: 5.15.91-1-MANJARO arch: x86_64 bits: 64 Desktop: Xfce
    v: 4.18.1 Distro: Manjaro Linux
Machine:
  Type: Laptop System: SchenkerGmbH product: SCHENKER VIA 15 Pro v: Standard
    serial: <superuser required>
  Mobo: NB02 model: PF5NU1G v: Standard serial: <superuser required>
    UEFI: American Megatrends v: N.1.07.A06 date: 08/29/2022
Battery:
  ID-1: BAT0 charge: 39.8 Wh (47.0%) condition: 84.7/91.6 Wh (92.4%)
    volts: 11.1 min: 11.6
CPU:
  Info: 8-core AMD Ryzen 7 4800H with Radeon Graphics [MT MCP] speed (MHz):
    avg: 1392 min/max: 1400/2900
Graphics:
  Device-1: AMD Renoir driver: amdgpu v: kernel
  Device-2: Chicony HD Webcam type: USB driver: uvcvideo
  Display: x11 server: X.Org v: 21.1.6 driver: X: loaded: amdgpu
    unloaded: modesetting dri: radeonsi gpu: amdgpu resolution: 1920x1080~60Hz
  API: OpenGL v: 4.6 Mesa 22.3.4 renderer: AMD Radeon Graphics (renoir LLVM
    15.0.7 DRM 3.42 5.15.91-1-MANJARO)
Network:
  Device-1: Intel Wi-Fi 6 AX200 driver: iwlwifi
  Device-2: Realtek RTL8111/8168/8411 PCI Express Gigabit Ethernet
    driver: r8169
```

I just uninstalled due to missing fan control when I blacklist `tuxedo_io`

**Comment by xPathin** [comment]
Hi @tuxedoxt any status on this issue?

**Comment by mt2506** [comment]
Hi same issue on TUXEDO Polaris 17 AMD Gen1.

**Comment by matt-hires** [comment]
It seems that kernel version `5.15.93-1` fixes this problem on my machine.

**Comment by tuxedoxt** [comment]
Yes the initial suspicion proved to be true. It comes down to HW IO communication (read/writes to EC) in combination with keyboard use... for certain devices. Possibly a FW issue, will investigate further.

For now focused on a solution, in the meantime some changes to the driver has proved to help. In general less communication with the hardware. The upcoming version of tuxedo-keyboard (v3.1.4) will include these tweaks.

I'll write when it's in the package repos, then we can collect feedback.

**Comment by vmaurin** [comment]
On Arch `tuxedo-keyboard-dkms` seems to have been updated already to 3.1.4, but it feels a bit like a regression from 3.1.3 on typing. I am running without tccd since the bug, and here, still without tccd, some key inputs seems to disappear now

**Comment by alexbradd** [comment]
Tested version 3.1.4 of tuxedo-keyboard on Arch linux with kernel version 6.1.12 on a Pulse 15 gen 2. Most of the lag is gone and is definitely better, however sometimes some keystrokes are still missed/repeated.

**Comment by xPathin** [comment]
Tested 3.1.4 and can confirm that the experience is a billion times better. At least the keyboard is usable again.

**Comment by riscygeek** [comment]
Can confirm on the TUXEDO Polaris 17 (AMD) Gen1.
OS: Linux Mint 21.1
tuxedo-keyboard: 3.1.4
tuxedo-control-center: 1.2.5

When booting into OpenBSD (which has no support for this laptop), no delay is noticeable.
In the TTY on Linux Mint the latency is reduced, but is still annoying.

**Comment by matt-hires** [comment]
Tested again, delay is barely noticeable, but some keys are missed as already mentioned before.
